### PR TITLE
DSTU2 code fixes and extensions

### PR DESCRIPTION
**Pull request by nschwertner**

### CODE REVIEW

File: src/middlewares/patient.js [hunk 1]
```diff
@@ -3,18 +3,26 @@
 
     var keyFor = {
         "Observation": "subject",
-        "MedicationPrescription": "patient"
+        "MedicationPrescription": "patient",
+        "Encounter": "patient"
```
Comment (jmandel) [inline src/middlewares/patient.js]: In DSTU2  every resource with a patient has a search parameter called "patient" (sometimes it's just an alias for "subject") so we don't need any special logic -- we can just automatically add  a "patient"value whenever this middleware is invoked; keyFor can go away.

Comment (nschwertner) [inline src/middlewares/patient.js]: Is there a definitive list of the resources "with a patient" in DSTU2?

Comment (niquola) [inline src/middlewares/patient.js]: We can extract from FHIR metadata all refs to Patient

Comment (jmandel) [inline src/middlewares/patient.js]: But we technically don't even need to any more -- at least, not strictly.
We can just tack a patient search param on whenever this middleware is
active. I was imagining that SMART's client at least would expose two
instances of fhir.js client:

 * smart.api  --> has patient middleware disabled
 * smart.patient.api --> has patient middleware enabled, for the in-context
patient.

Maybe this is over-thinking things, but it's convenient not to require an
explicit "patient" parameter all over the place.

then users can call whichever one they want in a given context.

Comment (nschwertner) [inline src/middlewares/patient.js]: Josh, sure, we can expose two instances of the client as you outlined with and without the patient context middleware enabled. Furthermore, the client can decide which of the two client to instantiate and expose based on the launch conditions (if there is a patient context). This way, an app which expects to work in patient context will not produce incorrect results by accident.

Still, this does not resolve the need of the patient context middleware to have a list of the resources for which it should apply the patient context (unless there is something that I don't yet understand here).

Comment (jmandel) [inline src/middlewares/patient.js]: All I'm saying is that *technically* providing a "patient=123" parameter on
arbitrary search operations should be okay. Any time there's a patient
associated with the resource, the search param will be correctly called
"patient". And any time there isn't, it should be ignored (at least, that's
FHIR's intended behavior for unrecognized query parameters). Of course,
having a list of resources would be nice, so we don't say "GET
/ValueSet?patient=123" -- but then again, a developer using out client
shouldn't be searching for value sets with the patient-context-embedded
object; the should use the patient-context-free object for that kind of
search.

**Comment by nschwertner** [comment]
Umm... The CI test for the NodeJS adapter hangs up. Not sure why though. @niquola could you take a look?

File: src/middlewares/patient.js (hunk 1 above)
Comment (nschwertner) [inline src/middlewares/patient.js]: Good. Let's do both then - bind variants of the client to`smart.api` and `smart.patient.api` in the SMART JS Client and have the sanity check list in `fhir.js` (granted, it's something else to maintain, but makes things a bit nicer from a user perspective)

**Comment by niquola** [comment]
Fixed

**Comment by nschwertner** [comment]
CI is still failing (see above)

**Pull request merged**